crates/services/executor/src/executor.rs

```diff
@@ -1030,228 +1080,671 @@ where
+            block_st_transaction,
+        )?;
+        let Input::Contract(input) = core::mem::take(&mut inputs[0]) else {
+            unreachable!()
+        };
+        let Output::Contract(output) = outputs[0] else {
+            unreachable!()
+        };
+        Ok((input, output))
+    }
+
+    fn check_mint_amount(mint: &Mint, expected_amount: u64) -> ExecutorResult<()> {
+        if *mint.mint_amount() != expected_amount {
+            return Err(ExecutorError::CoinbaseAmountMismatch)
+        }
+        Ok(())
+    }
+
+    fn check_gas_price(mint: &Mint, expected_gas_price: Word) -> ExecutorResult<()> {
+        if *mint.gas_price() != expected_gas_price {
+            return Err(ExecutorError::CoinbaseGasPriceMismatch)
+        }
+        Ok(())
+    }
+
+    fn check_mint_has_expected_index(
+        checked_mint: &Checked<Mint>,
+        execution_data: &ExecutionData,
+    ) -> ExecutorResult<()> {
+        if checked_mint.transaction().tx_pointer().tx_index() != execution_data.tx_count {
+            return Err(ExecutorError::MintHasUnexpectedIndex)
+        }
+        Ok(())
+    }
+
+    fn verify_mint_for_empty_contract(mint: &Mint) -> ExecutorResult<()> {
+        if *mint.mint_amount() != 0 {
+            return Err(ExecutorError::CoinbaseAmountMismatch)
+        }
+
+        let input = input::contract::Contract {
+            utxo_id: UtxoId::new(Bytes32::zeroed(), 0),
+            balance_root: Bytes32::zeroed(),
+            state_root: Bytes32::zeroed(),
+            tx_pointer: TxPointer::new(BlockHeight::new(0), 0),
+            contract_id: ContractId::zeroed(),
+        };
+        let output = output::contract::Contract {
+            input_index: 0,
+            balance_root: Bytes32::zeroed(),
+            state_root: Bytes32::zeroed(),
+        };
+        if mint.input_contract() != &input || mint.output_contract() != &output {
+            return Err(ExecutorError::MintMismatch)
+        }
+        Ok(())
+    }
+
+    fn store_mint_tx<T>(
+        mint: Mint,
+        execution_data: &mut ExecutionData,
+        coinbase_id: TxId,
+        block_st_transaction: &mut StorageTransaction<T>,
+    ) -> ExecutorResult<Transaction>
+    where
+        T: KeyValueInspect<Column = Column>,
+    {
+        let tx = mint.into();
+
+        execution_data.tx_status.push(TransactionExecutionStatus {
+            id: coinbase_id,
+            result: TransactionExecutionResult::Success {
+                result: None,
+                receipts: vec![],
+                total_gas: 0,
+                total_fee: 0,
+            },
+        });
+
+        if block_st_transaction
+            .storage::<ProcessedTransactions>()
+            .insert(&coinbase_id, &())?
+            .is_some()
+        {
+            return Err(ExecutorError::TransactionIdCollision(coinbase_id))
+        }
+        Ok(tx)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn execute_chargeable_transaction<Tx, T>(
```

Comment:
The `execute_chargeable_transaction` and `validate_chargeable_transaction` should be the same, except the `Self::validate_tx_outcome`. But as I mentioned before, you can move this check to `validate_block` level.

In will remove code duplication

Reply:
There is also this section from the old code:
```
        if execution_kind == ExecutionKind::Validation {
            self.validate_inputs_state(
                checked_tx.transaction().inputs(),
                tx_id,
                tx_st_transaction,
            )?;
        }
```
Is that not needed? I'm just copying what we had before, but happy to remove unneeded stuff.

Reply:
The comparison of the transactions already includes this check. We can still call this function after the tx was executed in the VM, just like a double check that all fields set by the block producer are correct. It applies to validation and execution.